src/ModelOrderReduction/component/forcefield/HyperReducedTriangleFEMForceField.inl

```diff
@@ -373,30 +374,57 @@ void HyperReducedTriangleFEMForceField<DataTypes>::draw(const core::visual::Visu
 #endif /* SOFA_NO_OPENGL */
 }
 
-
-template<class DataTypes>
-void HyperReducedTriangleFEMForceField<DataTypes>::addKToMatrix(sofa::linearalgebra::BaseMatrix *mat, SReal k, unsigned int &offset)
+template <class DataTypes>
+void HyperReducedTriangleFEMForceField<DataTypes>::buildStiffnessMatrix(core::behavior::StiffnessMatrix* matrix)
 {
-    if (d_performECSW.getValue())
-    {
-        int iECSW;
-        for(unsigned i = 0 ; i<m_RIDsize ;++i)
-        {
-            iECSW = reducedIntegrationDomain(i);
-            StiffnessMatrix JKJt,RJKJtRt;
-            this->computeElementStiffnessMatrix(JKJt, RJKJtRt, _materialsStiffnesses[iECSW], _strainDisplacements[iECSW], _rotations[iECSW]);
-            this->addToMatrix(mat,offset,(*_indexedElements)[iECSW],weights(iECSW)*RJKJtRt,-k);
+    StiffnessMatrix JKJt, RJKJtRt;
+    sofa::type::Mat<3, 3, Real> localMatrix(type::NOINIT);
 
-        }
-    }
+    constexpr auto S = DataTypes::deriv_total_size; // size of node blocks
+    constexpr auto N = Element::size();
+
+    auto dfdx = matrix->getForceDerivativeIn(this->mstate)
+                    .withRespectToPositionsIn(this->mstate);
+
+    sofa::Size triangleId = 0;
+
+    typename VecElement::const_iterator it;
+    auto it0=_indexedElements->begin();
+    int nbElementsConsidered;
+    if (!d_performECSW.getValue())
+        nbElementsConsidered = _indexedElements->size();
     else
+        nbElementsConsidered = m_RIDsize;
+
+    const bool performECSW = d_performECSW.getValue();
+    for( unsigned int numElem = 0 ; numElem<nbElementsConsidered ;++numElem)
     {
-        for(unsigned i=0; i< _indexedElements->size() ; i++)
+        if (!performECSW)
+        {
+            triangleId = numElem;
+        }
+        else
         {
-            StiffnessMatrix JKJt,RJKJtRt;
-            this->computeElementStiffnessMatrix(JKJt, RJKJtRt, _materialsStiffnesses[i], _strainDisplacements[i], _rotations[i]);
-            this->addToMatrix(mat,offset,(*_indexedElements)[i],RJKJtRt,-k);
+            triangleId = reducedIntegrationDomain(numElem);
+        }
+        it = it0 + triangleId;
+
+
+        this->computeElementStiffnessMatrix(JKJt, RJKJtRt, _materialsStiffnesses[triangleId], _strainDisplacements[triangleId], _rotations[triangleId]);
+
+        for (sofa::Index n1 = 0; n1 < N; n1++)
+        {
+            for (sofa::Index n2 = 0; n2 < N; n2++)
+            {
+                RJKJtRt.getsub(S * n1, S * n2, localMatrix); //extract the submatrix corresponding to the coupling of nodes n1 and n2
+                if (!d_performECSW.getValue())
```

Comment:
```suggestion
                if (!performECSW)
```

Reply:
Looking really into it now I think there are a lot of place where I can do that.
It's only to reduce the cost for the loop, right ?

Reply:
Yes, `.getValue()` is not free. It does not cost a lot, but if you accumulate the cost in a loop, it can become significant.

Reply:
Since it's not in the scope of the PR, I recommend to merge as it is. And if you want to extract the `getValue()` out of loops, you can do it in another PR.